docs/architecture/adr-022-multi-versioned-state-machine.md

```diff
@@ -0,0 +1,98 @@
+
+### SDK Context
+
+The `sdk.Context` type is passed to almost every component in the application and is thus ideal for storing and accessing the current app version. It already contains the app version in the block header. However this is not always populated. Thus `InitChain` and `PrepareProposal` were modified to include the app version as either passed from the genesis file or loaded from the application itself.
+
+### Module Manager Changes
+
+The module manager was modified to contain a mapping of modules to their version.
+
+```go
+type Manager struct {
+    versionedModules map[uint64]map[string]sdkmodule.AppModule
+    ...
+}
+```
+
+Each module has a `Name` and a `ConsensusVersion` that marks it unique. No two modules can have the same `Name` within a single app version. When constructing the application, we specify the contiguous range (inclusive) of versions that the module is part of:
+
+```go
+app.mm, err = module.NewManager([]module.VersionedModule{
+    {
+        Module:      blobstream.NewAppModule(appCodec, app.BlobstreamKeeper),
+        FromVersion: v1, ToVersion: v2,
+    },
+    {
+        Module:      upgrade.NewAppModule(app.UpgradeKeeper),
+        FromVersion: v2, ToVersion: v2,
+    },
+})
+```
+
+`InitGenesis`, `BeginBlock` and `EndBlock` will only be called for the modules that belong to the current version of the application.
+
+`DeliverTx` is not called by the `module.Manager`. To ensure that only relevant transactions are executed, the `MsgServiceRouter` had to be modified to reject the execution of transaction that are not part of the app version through the `CircuitBreaker` struct within it.
+
+### Configurator Changes
+
+The configurator is an object that is used upon initialisation to register the msg server, query server and migrations. In order to know what `sdk.Msg`s corresponded with which app version, a wrapper was added to the configurator that would scrape all the `sdk.Msg`s from the msg server as they were being registered and add them to a map:
+
+```go
+// acceptedMsgs is a map from appVersion -> msgTypeURL -> struct{}.
+acceptedMessages map[uint64]map[string]struct{}
+```
+
+### Transaction Filtering through the `MsgVersioningGateKeeper`
+
+The messages within a transactions are filtered based on the app version. The map constructed from the configurator is passed to a new AnteHandler which will reject transactions of a different app version in `CheckTx` so that clients get an immediate response. The `MsgVersioningGateKeeper` is also passed to the `MsgServiceRouter` to reject transactions upon execution. The reason for needing both is that another module may try to execute a transaction for a module that is not part of the app version.
```

Comment:
```suggestion
The messages within a transaction are filtered based on the app version. The map constructed from the configurator is passed to a new AnteHandler which will reject transactions of a different app version in `CheckTx` so that clients get an immediate response. The `MsgVersioningGateKeeper` is also passed to the `MsgServiceRouter` to reject transactions upon execution. The reason for needing both is that another module may try to execute a transaction for a module that is not part of the app version.
```